roadmap/implementers-guide/src/runtime/router.md

```diff
@@ -27,6 +36,61 @@ No initialization routine runs for this module.
 
 ## Routines
 
+There are situations when actions that took place within the relay chain could lead to a downward message
+sent to a para. For example, if an entry-point to transfer some funds to a para was called.
+
+For these cases, there are two routines, `has_dmq_capacity_for_relay_chain` and `send_downward_messages`,
+intended for use by the relay chain.
+
+`send_downward_messages` is used for enqueuing one or more downward messages for a certain recipient. Since downward
+message queues can hold only so many messages per one sender (and the relay chain is not an exception),
```

Comment:
Why should the Relay chain not be an exception? It is the Relay chain which must store these messages, so I don't see why it shouldn't be able to determine when it is acceptable to increase the buffer. Furthermore by limiting this, then other actions which need to be infallible (such as fund transfer notifications which happen on a scheduled block) may become impossible or result in the need for additional unbounded queues *also living on the relay chain*. Better, surely, to keep things simple and just allow the relay chain to add messages into the queue as it sees fit.

Reply:
Funnily, one of the reasons for this limitation was actually simplicity. Since there is, atm unbounded, set of downward message kinds it seemed to me the simplest to just slap a single limit on a downward message queue (DMQ) instead of coming up with individual mechanisms that prevent DoS for each operation that could lead to a downward message (DM) send. 

For example, one of such operations, is sending some funds into a parachain account which results in a DM. We of course trust the relay chain, but this particular operation can be executed by a signed extrinsic and therefore can be abused by sending dust transfers. Since HRMP piggybacks on DMQ in this proposal, regular users of the relay-chain would be able to compete with the HRMP messages from other paras. Users would have had an advantage since HRMP is limited by 1 message per candidate. Surely, there should be some sort of mitigation in place to prevent that. 

We could go with the ad-hoc mitigations. E.g. for transfer into para there is a simple mitigation: we can coalesce all incoming transfers per block into a single DM. Perhaps we would have to omit the senders for the sake of avoiding inflating the DM's size, but that might limit the usefulness of such message somewhat. For each future operation that might need to send a DM we would need to come up with a mitigation as well, and it might not be that simple.

The alternative is just slap a limit on each DMQ, if there is no space to send a message just bail and call it a day. As you noticed, that would rule out a certain kind of operations, specifically those which cannot check the capacity beforehand. So it is not just simple, it's too simplistic to the degree that's it is also too limiting.

Therefore, it seems that we have to bite the bullet and go with limitless relay-chain and ad-hoc mitigations.

Reply:
Yeah, basically for non-essential messages then I think aggregation or economic disincentives will need to be put in place.

HRMP should necessarily be expensive; it's mostly there to give a feel for the (XCMP) product before the product is actually ready. It's not really meant for production usage.